src/query_graph/graph_path.rs

```diff
@@ -2059,12 +2060,109 @@ impl OpGraphPath {
+            // sometime lead to a large number of additional paths to explore, which can have a substantial cost. So we want
+            // to limit it if we can avoid it. As it happens, we should return `true` if it is possible that "something"
+            // (some field) in the type of `field` is reachable in _another_ subgraph but no in the one of the current path.
+            // And while it's not trivial to check this in general, there are some easy cases we can eliminate. For instance,
+            // if the type in the current subgraph has only leaf fields, we can check that all other subgraphs reachable
+            // from the implementation have the same set of leaf fields.
+            let base_ty_name = field.ty.inner_named_type();
+            if is_leaf_type(schema, base_ty_name) {
+                continue;
+            }
+            let Some(ty) = schema.get_object(base_ty_name) else {
+                return Ok(true);
+            };
+            if ty
+                .fields
+                .values()
+                .any(|f| !is_leaf_type(schema, f.ty.inner_named_type()))
+            {
+                return Ok(true);
+            }
+            for node in self.graph.nodes_for_type(&ty.name) {
+                let node = self.graph.node_weight(node)?;
+                let tail = self.graph.node_weight(self.tail)?;
+                if node.source == tail.source {
+                    continue;
+                }
+                let Some(src) = self.graph.sources.get(&node.source) else {
+                    return Err(FederationError::internal(format!(
+                        "{node} has no valid schema in QueryGraph: {:?}",
+                        self.graph
+                    )));
+                };
+                let fed_spec = get_federation_spec_definition_from_subgraph(src)?;
+                let shareable_directive = fed_spec.shareable_directive(src)?;
+                let build_err = || {
+                    Err(FederationError::internal(format!(
+                        "{implem} is an object in {} but a {} in {}",
+                        tail.source, node.type_, node.source
+                    )))
+                };
+                let QueryGraphNodeType::SchemaType(node_ty) = &node.type_ else {
+                    return build_err();
+                };
+                let node_ty = node_ty.get(schema)?;
+                let other_fields = match node_ty {
+                    ExtendedType::Object(obj) => &obj.fields,
+                    ExtendedType::Interface(int) => &int.fields,
+                    _ => return build_err(),
+                };
+                let Some(field) = other_fields.get(&itf.field_name) else {
+                    continue;
+                };
+                if field
+                    .directives
+                    .iter()
+                    .any(|d| d.name == shareable_directive.name)
+                {
```

Comment:
This looks opposite from the JS:
```ts
        if (!fieldInOther || !fieldInOther.hasAppliedDirective(otherMetadata.shareableDirective())) {
          // The shareable field is actually not shared here (it's either not declared, or external), so we can ignore that subgraph.
          continue;
        }
```
```suggestion
                if !field.directives.has(&shareable_directive.name) {
```